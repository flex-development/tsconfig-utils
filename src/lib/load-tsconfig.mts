/**
 * @file loadTsconfig
 * @module tsconfig-utils/lib/loadTsconfig
 */

import normalizeRelativePaths from '#internal/normalize-relative-paths'
import mergeTsconfig from '#lib/merge-tsconfig'
import readTsconfig from '#lib/read-tsconfig'
import type { ModuleId } from '@flex-development/mlly'
import * as mlly from '@flex-development/mlly'
import type {
  LoadTsconfigOptions,
  ResolvedTsconfig,
  Tsconfig
} from '@flex-development/tsconfig-utils'
import { ksort, omit, shake, trim } from '@flex-development/tutils'

/**
 * Load a tsconfig file.
 *
 * @see {@linkcode ModuleId}
 * @see {@linkcode LoadTsconfigOptions}
 * @see {@linkcode ResolvedTsconfig}
 *
 * @async
 *
 * @this {void}
 *
 * @param {ModuleId} id
 *  Module id of tsconfig file, or path to tsconfig file
 * @param {LoadTsconfigOptions | null | undefined} [options]
 *  Load options
 * @return {Promise<ResolvedTsconfig | null>}
 *  Resolved tsconfig or `null` if tsconfig file is not found
 */
async function loadTsconfig(
  this: void,
  id: ModuleId,
  options?: LoadTsconfigOptions | null | undefined
): Promise<ResolvedTsconfig | null> {
  /**
   * Resolved tsconfig object.
   *
   * @var {ResolvedTsconfig | null} tsconfig
   */
  let resolved: ResolvedTsconfig | null = await readTsconfig(id, options)

  if (resolved !== null) {
    /**
     * List of property paths where the value may be a relative path.
     *
     * @const {Set<string>} relativePaths
     */
    const relativePaths: Set<string> = new Set(options?.relativePaths ?? [
      'compilerOptions.baseUrl',
      'compilerOptions.outDir',
      'compilerOptions.paths.*',
      'compilerOptions.paths.*.*',
      'compilerOptions.rootDir',
      'exclude.*',
      'files.*',
      'include.*'
    ])

    if (
      resolved.tsconfig.extends !== false &&
      resolved.tsconfig.extends !== null &&
      resolved.tsconfig.extends !== undefined
    ) {
      /**
       * Configuration files to inherit from.
       *
       * @const {string[]} bases
       */
      const bases: string[] = [resolved.tsconfig.extends]
        .flat()
        .map(trim)
        .filter(extend => extend.length > 0)

      for (const specifier of bases) {
        /**
         * Resolved base tsconfig.
         *
         * @const {ResolvedTsconfig | null} extend
         */
        const extend: ResolvedTsconfig | null = await loadTsconfig(
          await mlly.resolveModule(specifier, resolved.url, options),
          options
        )

        if (extend) {
          normalizeRelativePaths(
            resolved.url,
            extend.url,
            extend.tsconfig,
            null,
            relativePaths
          )

          /**
           * Tsconfig to inherit from.
           *
           * @const {Tsconfig} base
           */
          const base: Tsconfig = omit(extend.tsconfig, ['references'])

          resolved.tsconfig = mergeTsconfig(base, resolved.tsconfig)
        }
      }

      delete resolved.tsconfig.extends
    }

    normalizeRelativePaths(
      resolved.url,
      resolved.url,
      resolved.tsconfig,
      null,
      relativePaths
    )

    return {
      tsconfig: ksort(shake(resolved.tsconfig), { deep: true }),
      url: resolved.url
    }
  }

  return null
}

export default loadTsconfig
