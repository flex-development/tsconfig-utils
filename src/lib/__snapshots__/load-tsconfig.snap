// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/loadTsconfig > should return resolved tsconfig object 1`] = `
{
  "tsconfig": {
    "$schema": "https://json.schemastore.org/tsconfig",
    "_version": "2.0.0",
    "compilerOptions": {
      "allowImportingTsExtensions": true,
      "allowJs": true,
      "allowUnreachableCode": false,
      "allowUnusedLabels": false,
      "alwaysStrict": true,
      "checkJs": true,
      "customConditions": [
        "tsconfig-utils",
        "development",
      ],
      "declaration": true,
      "declarationMap": true,
      "esModuleInterop": true,
      "exactOptionalPropertyTypes": true,
      "isolatedModules": true,
      "lib": [
        "es2023",
      ],
      "module": "esnext",
      "moduleResolution": "bundler",
      "newLine": "lf",
      "noEmit": true,
      "noErrorTruncation": true,
      "noFallthroughCasesInSwitch": true,
      "noImplicitOverride": true,
      "noImplicitReturns": true,
      "noPropertyAccessFromIndexSignature": true,
      "noUncheckedIndexedAccess": true,
      "noUncheckedSideEffectImports": true,
      "noUnusedLocals": true,
      "noUnusedParameters": false,
      "paths": {
        "https://esm.sh/lodash@4.17.21": [
          "./node_modules/@types/lodash/index.d.ts",
        ],
      },
      "resolveJsonModule": true,
      "skipLibCheck": true,
      "sourceMap": true,
      "strict": true,
      "target": "esnext",
      "verbatimModuleSyntax": true,
    },
    "exclude": [
      "coverage",
      "dist",
      "node_modules",
    ],
    "include": [
      "**/**.json",
      "**/**.mjs",
      "**/**.mts",
      "**/**.ts",
    ],
    "mdx": {
      "plugins": [
        "remark-directive",
        "remark-frontmatter",
        "remark-gfm",
      ],
    },
  },
  "url": file://\${process.cwd()}/__fixtures__/tsconfig.json,
}
`;
